pylint/checkers/variables.py

```diff
@@ -976,267 +989,258 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
+    )
+    def _undefined_and_used_before_checker(
+        self, node: nodes.Name, stmt: nodes.NodeNG
+    ) -> None:
         frame = stmt.scope()
         start_index = len(self._to_consume) - 1
 
-        undefined_variable_is_enabled = self.linter.is_message_enabled(
+        self._undefined_variable_is_enabled = self.linter.is_message_enabled(
             "undefined-variable"
         )
-        used_before_assignment_is_enabled = self.linter.is_message_enabled(
+        self._used_before_assignment_is_enabled = self.linter.is_message_enabled(
             "used-before-assignment"
         )
 
         # iterates through parent scopes, from the inner to the outer
         base_scope_type = self._to_consume[start_index].scope_type
-        # pylint: disable=too-many-nested-blocks; refactoring this block is a pain.
+
         for i in range(start_index, -1, -1):
             current_consumer = self._to_consume[i]
 
-            # The list of base classes in the class definition is not part
-            # of the class body.
-            # If the current scope is a class scope but it's not the inner
-            # scope, ignore it. This prevents to access this scope instead of
-            # the globals one in function members when there are some common
-            # names.
-            if current_consumer.scope_type == "class" and (
-                utils.is_ancestor_name(current_consumer.node, node)
-                or (i != start_index and self._ignore_class_scope(node))
-            ):
+            # Certain nodes shouldn't be checked as they get checked another time
+            if self._node_should_be_skipped(current_consumer, node, i == start_index):
                 continue
 
-            # Ignore inner class scope for keywords in class definition
-            if (
-                current_consumer.scope_type == "class"
-                and isinstance(node.parent, nodes.Keyword)
-                and isinstance(node.parent.parent, nodes.ClassDef)
-            ):
-                continue
+            action, found_nodes = self._check_consumer(
+                node, stmt, frame, current_consumer, i, base_scope_type
+            )
 
-            # if the name node is used as a function default argument's value or as
-            # a decorator, then start from the parent frame of the function instead
-            # of the function frame - and thus open an inner class scope
-            if (
-                current_consumer.scope_type == "function"
-                and self._defined_in_function_definition(node, current_consumer.node)
-            ):
-                # ignore function scope if is an annotation/default/decorator, as not in the body
+            if not action.value:
                 continue
+            if action.value == 2:
+                current_consumer.mark_as_consumed(node.name, found_nodes)
+            if action.value:
+                return
```

Comment:
```suggestion
            if action == VariableVisitConsumerAction.CONTINUE:
                continue
            if action == VariableVisitConsumerAction.CONSUME:
                current_consumer.mark_as_consumed(node.name, found_nodes)
            if action == VariableVisitConsumerAction.RETURN:
                return
```
This makes errors less likely, if the code was wrong it would look very wrong here. Also you do not have to know about the underlying value at all.

Reply:
The last row doesn't work as `CONSUME` should also return. Should we do `action in {VariableVisitConsumerAction.RETURN, VariableVisitConsumerAction.CONSUME}` or keep `if action.value `?

Reply:
I would do the former, it's more explicit as evidenced by the fact I just misunderstood what 'action.value' was doing :smile: Also it does not rely on the value of the other enum being falsey.